### DISCUSSION TITLE
How to set-output from a python script in Github workflow step.

### DISCUSSION THREAD

**Discussion by suhail-ali-m** [community/community#28146] [Actions] [answered]
`run: echo "::set-output name=secondword::world"`

I need to achieve the same from a python script.
` -name: Parse the html

        run: |
          import os
          paths="${{ steps.validat_test_file_exist.outputs.TEST_REPORTS }}"
          files=paths.strip().split(' ')
          result = ""
          for path in files:
            print(path)
            print(os.path.exists(path))
            with open(path) as fp:
              module=path.split('/')[0].title() + " Module: "
              result+=module
          echo "::set-output name=test_report::result"
        shell: python
`

Here I want to output the "result" of python script so that it can be used by other jobs/workflows.
Is it only possible to set-output from a bash script?

**Comment by airtower-luna** [comment, marked as answer]
`set-output` can be used from anything that writes to standard output. In Python the easiest solution is to use `print()`:

```py
print(f'::set-output name=test_report::{result}')
```

Happy Python'ing! :snake:

**Reply by suhail-ali-m** [reply]
I achieved the same with subprocess.
`cmd='echo \"::set-output name= test_report::'+result+'\"'`
`subprocess.call([str(cmd)], shell=True)`

Thanks, this is great.  I will replace it with your solution.

**Reply by ahatzz11** [reply]
@airtower-luna Now that `::set-output` is being deprecated, do you have any tips on doing this with the `GITHUB_OUTPUT` environment variable?

Blog: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

**Reply by airtower-luna** [reply]
Do the same thing as with the shell, *append* to the file specified in the `GITHUB_OUTPUT` environment variable:

```py
import os
name = 'my_name'
value = 'my_value'
with open(os.environ['GITHUB_OUTPUT'], 'a') as fh:
    print(f'{name}={value}', file=fh)
```

Mind that for values with multiple lines you now have to use the same format as for [multiline environment variables](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings), instead of URL encoding them.

**Reply by Bullrich** [reply]
For those struggling to do the multiline output work, I made a small method that set its up for you:
```python
import uuid

def set_multiline_output(name, value):
    with open(os.environ['GITHUB_OUTPUT'], 'a') as fh:
        delimiter = uuid.uuid1()
        print(f'{name}<<{delimiter}', file=fh)
        print(value, file=fh)
        print(delimiter, file=fh)
```

Just call it with:
```python
set_multiline_output("value", my_multiline_string)
```

**Reply by baslia** [reply]
Thanks !
I was looking for an easy way to do this !

**Reply by kinguser981** [reply]
i used this method:

`
python3 convertor.py ${{ github.event.inputs.FILE_LINK }} > link.txt
`
`
echo "FILE_ID=$(<link.txt)" >> $GITHUB_ENV
`

usage: ${{ env.FILE_ID }}

**Comment by Bullrich** [comment]
Providing a new updated answer with methods:
```python
import uuid

def set_output(name, value):
    with open(os.environ['GITHUB_OUTPUT'], 'a') as fh:
        print(f'{name}={value}', file=fh)


def set_multiline_output(name, value):
    with open(os.environ['GITHUB_OUTPUT'], 'a') as fh:
        delimiter = uuid.uuid1()
        print(f'{name}<<{delimiter}', file=fh)
        print(value, file=fh)
        print(delimiter, file=fh)
```

This two methods should do exactly what you need.

**Comment by cpfarherFinitestate** [comment]
Hello, I'm trying to create a github action in the markeplace. I defined an output whose name is "response" in the `action.yml` and I have a python script that mainly is called in a Dockerfile with: `ENTRYPOINT [ "python3","myapp.py"]`
I tried to follow this solution: https://github.com/orgs/community/discussions/28146#discussioncomment-4110404 and when I tried to get from {steps.id.outputs.response} I'm getting and empty value. Do you have any idea why this happen?  Also the deprecated way does not work. :(

**Reply by airtower-luna** [reply]
In theory it should work just the same. Hard to say what might be going wrong without seeing your action & test workflow. :thinking:

**Reply by cpfarherFinitestate** [reply]
Sorry, typo, NVM, I solved it.
